packages/react/src/ui-react/hooks/use-async-effect.ts

```diff
@@ -32,16 +32,25 @@ export type AsyncEffectCallback = () => Promise<unknown>;
  */
 export function useAsyncEffect(
     effect: AsyncEffectCallback,
-    deps?: React.DependencyList
+    deps?: DependencyList
 ): void {
-    React.useEffect(
+    useEffect(
         () => {
             const ret = effect();
-            if (!isPromiseLike(ret) && typeof ret === "function") {
-                // Cleanup functions should be returned
+
+            // If the returned value is not an async function, just treat it
+            // like useEffect
+            if (typeof ret === "function") {
                 return ret;
             }
-            // Void otherwise
+
+            if (isPromiseLike(ret)) {
+                ret.then((cleanup) => {
+                    if (typeof cleanup === "function") {
```

Comment:
Instead of checking at runtime here, can we change the type of AsyncEffectCallback to get better type checking?

I think this may work:

```
export type AsyncEffectCallback = () =>
    | void
    | (() => void)
    | Promise<void>
    | Promise<() => void>;
```

Or maybe better, we could forget about handling the non-promise case (because why are you using useAsyncEffect if it doesn't return a promise?), and simplify to:

```
export type AsyncEffectCallback = () => Promise<void | (() => void)>;
```

Not sure why I had <unknown> in the type before. Probably no good reason.

Reply:
If we do the latter we can also remove the isPromiseLike check since it's always going to be a promise or undefined.